File: 19.删除链表的倒数第-n-个结点.cpp
/*
 * @lc app=leetcode.cn id=19 lang=cpp
 *
 * [19] 删除链表的倒数第 N 个结点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    public:
        ListNode* removeNthFromEnd(ListNode* head, int n) {
            ListNode* dummyHead = new ListNode(0);
            dummyHead->next = head;
            int cnt = 0;
            while (head)
            {
                cnt ++;
                head = head->next;
            }
            ListNode* p = dummyHead;
            int diff = cnt - n;
            while (diff --)
            {
                p = p->next;
            }
            if (p->next)
            {
                p->next = p->next->next;
            }
            else return nullptr;
            return dummyHead->next;
        }
    };
// @lc code=end
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    public:
        ListNode* removeNthFromEnd(ListNode* head, int n) {
            ListNode* dummyHead = new ListNode(0);
            dummyHead->next = head;
            ListNode* slow = dummyHead;
            ListNode* fast = dummyHead;
            while (n -- && fast)
            {
                fast = fast->next;
            }
            fast = fast->next;
            while (fast)
            {
                slow = slow->next, fast = fast->next;
            }
            slow->next = slow->next->next;
            return dummyHead->next;
            
        }
    };
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    public:
        ListNode* removeNthFromEnd(ListNode* head, int n) {
            ListNode* dummyHead = new ListNode(0);
            dummyHead->next = head;
            ListNode* slow = dummyHead;
            ListNode* fast = dummyHead;
            int step = n + 1;
            while (step --)
            {
                fast = fast->next;
                if (fast == nullptr) break;
            }
            while (fast)
            {
                slow = slow->next, fast = fast->next;
            }
            if (slow->next)
            {   
                slow->next = slow->next->next;
            }
            return dummyHead->next;
            
        }
    };

File: HOT100/HOT100p1/7.2/19.删除链表的倒数第-n-个结点.py
#
# @lc app=leetcode.cn id=19 lang=python3
#
# [19] 删除链表的倒数第 N 个结点
#
from mytools import *
# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        p1 = p2 = dummy
        for _ in range(n): p2 = p2.next
        while p2.next:
            p2 = p2.next
            p1 = p1.next
        p1.next = p1.next.next
        return dummy.next
# @lc code=end

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        p1, p2, dummy = ListNode(), ListNode(), ListNode()
        p1.next = head
        p2.next = head
        dummy.next = head
        while n:
            p2 = p2.next
            n -= 1
        f = True
        while p2.next:
            f = False
            p1 = p1.next
            p2 = p2.next
        p1.next = p1.next.next
        return p1.next if f else dummy.next